### ISSUE TITLE
Add Adj functor to AssertQubit

### ISSUE THREAD

**Issue by msoeken** [microsoft/qsharp-runtime#316]
## Describe the bug
Not really a bug, but an inconsistency. `AssertQubit` does not have an `Adj` functor, but `AssertAllZero`, which uses `AssertQubit`, does have one (`adjoint self`).

## To Reproduce
Consider the following operation:

```Q#
operation ApplyAnd(control1 : Qubit, control2 : Qubit, target : Qubit) : Unit is Adj {
  AssertQubit(Zero, target);
  CCNOT(control1, control2, target);
}
```

This does not compile because `AssertQubit` does not have an `Adj` functor.

## Expected behavior
Operation should compile, and check in the adjoint call to `ApplyAnd` that `target` is released in state `Zero`.

## Actual behavior
Operation does not compile, however, I can use the following alternative:

```Q#
operation ApplyAnd(control1 : Qubit, control2 : Qubit, target : Qubit) : Unit is Adj {
  AssertAllZero([target]);
  CCNOT(control1, control2, target);
}
```

**Comment by samarsha** [comment]
Yes, it seems like all of the `Assert*` operations should support both adjoint and controlled - @cgranade?

**Comment by cgranade** [comment]
I agree with `Adjoint`, but `Controlled` is a bit more subtle. Assertions work by reasoning about the probability of hypothetical measurements, but then that measurement isn't consistent with `Controlled`. For example:
```
/// # Summary
/// Given a qubit in the |0⟩ state, prepares it in the |1⟩ state.
operation PrepareOne(target : Qubit) : Unit is Adj + Ctl {
    X(target);
    AssertQubit(One, target);
}

/// # Summary
/// Prepares a maximally entangled pair on two qubits,
/// given two qubits each initially in the |0⟩ state.
operation PrepareBellPair(left : Qubit, right : Qubit) : Unit is Adj + Ctl {
    H(left);
    // Does this assertion pass or fail?
    Controlled PrepareOne([left], right);
}
```

**Comment by swernli** [comment]
@cgranade Given your concern about the controlled functor for assertions, should we remove `Ctl` from `AssertMeasurement` and `AssertMeasurementProbability`? These both define a controlled behavior that just ignores the control qubits...

**Comment by swernli** [comment]
@bettinaheim, @cgranade, @msoeken, bringing this back up so we can land at a decision for the associated PR. As it stands, `AssertMeasurement` and `AssertMeasurementProbability` don't have `Adj` but do have `Ctl` and my PR just adds `Adj` but leaves any existing controlled support as-is. From the sound of the discussion here and in the PR, the controlled behavior (which just ignores the controls) seems wrong. Do we want to open a separate issue to remove the controlled support, since that would technically be a breaking change, or do we want to squeeze that into this same issue + PR?

**Comment by bettinaheim** [comment]
I believe this is resolved with #450
